spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -443,89 +480,148 @@ public void run() {
 			}
 		}
 
-		private void invokeListener(final ConsumerRecord<K, V> record) {
-			try {
-				if (this.acknowledgingMessageListener != null) {
-					this.acknowledgingMessageListener.onMessage(record, new Acknowledgment() { //NOSONAR - long anon.
-
-						@Override
-						public void acknowledge() {
-							if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL)) {
-								updateManualOffset(record);
-							}
-							else if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)
-									|| ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE_SYNC)) {
-								ackImmediate(record);
-							}
-							else {
-								throw new IllegalStateException("AckMode must be MANUAL, "
-										+ "MANUAL_IMMEDIATE, or MANUAL_IMMEDIATE_SYNC for manual acks");
-							}
-						}
-
-						private void ackImmediate(final ConsumerRecord<K, V> record) {
-							if (Thread.currentThread().equals(ListenerConsumer.this.consumerThread)) {
-								Map<TopicPartition, OffsetAndMetadata> commits = Collections.singletonMap(
-										new TopicPartition(record.topic(), record.partition()),
-										new OffsetAndMetadata(record.offset() + 1));
-								if (ListenerConsumer.this.logger.isDebugEnabled()) {
-									ListenerConsumer.this.logger.debug("Committing: " + commits);
-								}
-								if (ListenerConsumer.this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)) {
-									ListenerConsumer.this.consumer.commitAsync(commits,
-											ListenerConsumer.this.commitCallback);
-								}
-								else {
-									ListenerConsumer.this.consumer.commitSync(commits);
-								}
-							}
-							else {
-								throw new IllegalStateException(
-										"With " + ListenerConsumer.this.ackMode.name()
-										+ " ack mode, acknowledge() must be invoked on the consumer thread");
-							}
-						}
+		private ConsumerRecords<K, V> checkPause(ConsumerRecords<K, V> unsent) {
+			if (this.paused && this.recordsToProcess.size() < this.queueDepth) {
```

Comment:
Shouldn't you check if the unsent ConsumerRecords all fit in the queue? If not, the sendToListener could fail again and you loss the unsent records. Also, since you resumed it already, next poll would return new records you have to deal with. I would wait to resume until the unsent ConsumerRecords are all sent to the listener